lldb/source/Commands/CommandObjectDWIMPrint.cpp

```diff
@@ -170,6 +170,14 @@ void CommandObjectDWIMPrint::DoExecute(StringRef command,
     ExpressionResults expr_result = target.EvaluateExpression(
         expr, exe_scope, valobj_sp, eval_options, &fixed_expression);
 
+    auto persistent_name = valobj_sp->GetName();
+    // EvaluateExpression doesn't generate a new persistent result (`$0`) when
+    // the expression is already just a persistent variable (`$var`). Instead,
+    // the same persistent variable is reused. Take note of when a persistent
+    // result is created, to prevent unintentional deletion of a user's
+    // persistent variable.
+    bool did_persist_result = persistent_name != expr;
```

Comment:
This seems like a really fragile way to test what you want to test, which is whether the result of the expression was a new expression result variable.  It might be better to add a "PeekNextPersistentVariableName" that just prints what the next persistent variable would be called but doesn't increment the counter, then assert that the variable you produced is NOT called that?

Reply:
I agree that it's not ideal, but I felt it was sufficient.

> what you want to test, which is whether the result of the expression was a new expression result variable

exactly. However I think peeking at the result variables is also not ideal. Both my initial stab at this, and your suggestion are deducing whether a new expression result variable was created. If I'm to change the API, I think it'd be ideal to make the API communicate have means to communicate this explicitly.

There's another approach I considered taking. Before calling `EvaluateExpression`, try the expression as a persistent variable and check in the target to see if it exists, and if it does exist, use it (and avoid `EvaluateExpression` altogether). This is similar to how expressions are first treated as frame variables (and as @felipepiovezan has requested, we should try `target variable` too).

Reply:
It would be good to have EvaluateExpression tell you whether it created a result variable or not.  For instance, calling a void function produces no result.  It would be nice to know that happened directly rather than attempting to guess from the returned ValueObject.  We definitely know this...
That would definitely be a better way to do this.

Reply:
Jim, I have updated the PR to try the expression as a persistent variable – prior to expression evaluation. My thinking is, instead of evaluating first, then asking if a persistent variable was created, it's more straight to the point to first ask "is this a persistent variable".

I like the idea of trying the various kinds of variables in sequence: frame, target, persistent.